src/PrestaShopBundle/Form/Admin/Sell/Product/VirtualProductFileType.php

```diff
@@ -0,0 +1,117 @@
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Validator\Constraints\File;
+use Symfony\Component\Validator\Constraints\Length;
+use Symfony\Component\Validator\Constraints\NotBlank;
+
+class VirtualProductFileType extends TranslatorAwareType
+{
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $maxUploadSize = $this->getConfiguration()->get('PS_ATTACHMENT_MAXIMUM_SIZE') . 'M';
+
+        $builder
+            ->add('has_file', SwitchType::class, [
+                'label' => $this->trans('Does this product have an associated file?', 'Admin.Catalog.Feature'),
+            ])
+            ->add('virtual_product_file_id', HiddenType::class)
+            ->add('file', DownloadableFileType::class, [
+                'label' => $this->trans('File', 'Admin.Global'),
+                'file_options' => [
+                    'help' => $this->trans(
+                        'Upload a file from your computer (%maxUploadSize% max.)',
+                        'Admin.Catalog.Help',
+                        ['%maxUploadSize%' => $maxUploadSize]
+                    ),
+                    'constraints' => [
+                        new File(['maxSize' => $maxUploadSize]),
+                    ],
+                ],
+            ])
+            ->add('name', TextType::class, [
+                'label' => $this->trans('Filename', 'Admin.Global'),
+                'help' => $this->trans('The full filename with its extension (e.g. Book.pdf)', 'Admin.Catalog.Help'),
+                'constraints' => [
+                    new NotBlank(),
+                    new TypedRegex(TypedRegex::TYPE_GENERIC_NAME),
+                    new Length([
+                        'max' => VirtualProductFileSettings::MAX_DISPLAY_FILENAME_LENGTH,
+                    ]),
+                ],
+            ])
+            ->add('download_times_limit', NumberType::class, [
+                'label' => $this->trans('Number of allowed downloads', 'Admin.Catalog.Feature'),
+                'help' => $this->trans(
+                    'Number of downloads allowed per customer. Set to 0 for unlimited downloads.',
+                    'Admin.Catalog.Help'
+                ),
+                'required' => false,
+                'constraints' => [
+                    new Length([
+                        'max' => VirtualProductFileSettings::MAX_DOWNLOAD_TIMES_LIMIT_LENGTH,
+                    ]),
+                ],
+            ])
+            ->add('access_days_limit', NumberType::class, [
+                'label' => $this->trans('Number of days', 'Admin.Catalog.Feature'),
+                'help' => $this->trans(
+                    'Number of days this file can be accessed by customers. Set to zero for unlimited access.',
+                    'Admin.Catalog.Help'
+                ),
+                'required' => false,
+                'constraints' => [
+                    new Length([
```

Comment:
I don't believe that `Length` is a constraint adapted for integers, it's supposed to be used for string values to limit the number of characters

In the same spirit, in the legacy model you use this new setting const to replace the `limit` parameter of the `access_days_limit` field, but it's not a `varchar` either in database and limit don't have the same purpose for integer and string values.

For string it represents the number of character, and for integer it's the number of bytes, so I'm not sure it this needs to be extracted in a configuration const. But at least I believe using `Range` constraint or `lessThan` would be more adapted here.

Reply:
| I don't believe that Length is a constraint adapted for integers, it's supposed to be used for string values to limit the number of characters

well it does the same, doesn't it? I didn't dig too deep, but it was a "size" constraint ob object model, same as for all the strings, it should just not allow number like "10000000000" or so and itseems to work as expected.
![Screenshot from 2021-04-06 12-12-14](https://user-images.githubusercontent.com/31609858/113687896-74fcfc00-96d1-11eb-8660-a215b07040fb.png)

Hmm if its not ok, then if i get it right, you suggest me to use LessThan constraint with hardcoded value of "10000000000" ? (that would look really crazy :sweat: )

Reply:
or maybe I should change the constant in settings like `MAX_ACCESSIBLE_DAYS_LIMIT = 1000000000`, leave the object model unchanged and use this value in LessThan ?

Reply:
OOhh okay, I didn't think `size` had this meaning in legacy object, for an integer I thought it was just a way to define the database field like size 8 would mean it's an `INT(8)` I though it defined the database column but you're right it seems like the legacy object doesn't care about the field type 😅

Wether it's a string or an int it appears to be checking the number of characters.. But it's a bit weird to have a constraint on a integer related to its number of characters.. So yes I think we could improve this a bit and make it more consistent by defining some max values (as you suggested) instead of lax length, which is more adapted for integer fields

Reply:
And yes leave the object model whit its hard coded value I guess